src/EditorFeatures/Core/Implementation/CommentSelection/AbstractCommentSelectionCommandHandler.cs

```diff
@@ -0,0 +1,156 @@
+        {
+            if (!buffer.CanApplyChangeDocumentToWorkspace())
+            {
+                return VSCommanding.CommandState.Unspecified;
+            }
+
+            return VSCommanding.CommandState.Available;
+        }
+
+        protected static void Format(ICommentSelectionService service, ITextSnapshot snapshot, IEnumerable<CommentTrackingSpan> changes, CancellationToken cancellationToken)
+        {
+            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return;
+            }
+
+            // Only format uncomment actions.
+            var textSpans = changes
+                .Where(change => change.Operation == Operation.Uncomment)
+                .Select(uncommentChange => uncommentChange.ToSnapshotSpan(snapshot).Span.ToTextSpan())
+                .ToImmutableArray();
+            var newDocument = service.FormatAsync(document, textSpans, cancellationToken).WaitAndGetResult(cancellationToken);
+            newDocument.Project.Solution.Workspace.ApplyDocumentChanges(newDocument, cancellationToken);
+        }
+
+        protected static void InsertText(List<TextChange> textChanges, int position, string text)
+        {
+            textChanges.Add(new TextChange(new TextSpan(position, 0), text));
+        }
+
+        protected static void DeleteText(List<TextChange> textChanges, TextSpan span)
+        {
+            textChanges.Add(new TextChange(span, string.Empty));
+        }
+
+        internal bool ExecuteCommand(ITextView textView, ITextBuffer subjectBuffer, Operation operation, CommandExecutionContext context)
+        {
+            var title = GetTitle(operation);
+            var message = GetMessage(operation);
+
+            using (context.OperationContext.AddScope(allowCancellation: false, message))
+            {
+
+                var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                if (document == null)
+                {
+                    return true;
+                }
+
+                var service = GetService(document);
+                if (service == null)
+                {
+                    return true;
+                }
+
+                var trackingSpans = new List<CommentTrackingSpan>();
+                var textChanges = new List<TextChange>();
+                CollectEdits(
+                    document, service, textView.Selection.GetSnapshotSpansOnBuffer(subjectBuffer),
+                    textChanges, trackingSpans, operation, CancellationToken.None);
+
+                using (var transaction = new CaretPreservingEditTransaction(title, textView, _undoHistoryRegistry, _editorOperationsFactoryService))
+                {
+                    document.Project.Solution.Workspace.ApplyTextChanges(document.Id, textChanges.Distinct(), CancellationToken.None);
```

Comment:
1. if textchanges is empty, should we even do any of this work?
2. if text changes cannot be empty, can we document that somehow.